ext/MatrixAlgebraKitMooncakeExt/MatrixAlgebraKitMooncakeExt.jl

```diff
@@ -0,0 +1,225 @@
+                else
+                    map(A_ -> Mooncake.build_tangent(typeof(A_), real(A_), imag(A_)), dAtmp_)
+                end
+                Mooncake.increment!!(Mooncake.tangent(A_dA), dAtmp)
+                arg1 .= arg1c
+                arg2 .= arg2c
+                A .= Ac
+                darg1 .= zero(darg1)
+                darg2 .= zero(darg2)
+                return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+            end
+            return Mooncake.CoDual(args, dargs), $adj
+        end
+    end
+end
+
+for (f, f_full, pb, adj) in (
+        (qr_null!, qr_full, qr_null_pullback!, :dqr_null_adjoint),
+        (lq_null!, lq_full, lq_null_pullback!, :dlq_null_adjoint),
+    )
+    @eval begin
+        @is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof($f), Any, Any, MatrixAlgebraKit.AbstractAlgorithm}
+        function Mooncake.rrule!!(f_df::CoDual{typeof($f)}, A_dA::CoDual, arg_darg::CoDual, alg_dalg::CoDual{<:MatrixAlgebraKit.AbstractAlgorithm}; kwargs...)
+            A, dA = arrayify(A_dA)
+            Ac = copy(A)
+            arg, darg = arrayify(arg_darg)
+            argc = copy(arg)
+            # WHY is this copy needed?
+            arg = $f(copy(A), arg, Mooncake.primal(alg_dalg))
+            function $adj(::Mooncake.NoRData)
+                dAtmp_ = zero(dA)
+                dAtmp_ .= $pb(dAtmp_, A, arg, darg; kwargs...)
+                dAtmp = if eltype(dA) <: Real
+                    dAtmp_
+                else
+                    map(A_ -> Mooncake.build_tangent(typeof(A_), real(A_), imag(A_)), dAtmp_)
+                end
+                Mooncake.increment!!(Mooncake.tangent(A_dA), dAtmp)
+                A .= Ac
+                arg .= argc
+                darg .= zero(darg)
+                return Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData(), Mooncake.NoRData()
+            end
+            return arg_darg, $adj
+        end
+    end
+end
+
+@is_primitive Mooncake.DefaultCtx Mooncake.ReverseMode Tuple{typeof(MatrixAlgebraKit.eig_vals!), Any, Any, MatrixAlgebraKit.AbstractAlgorithm}
+function Mooncake.rrule!!(::CoDual{<:typeof(MatrixAlgebraKit.eig_vals!)}, A_dA::CoDual, D_dD::CoDual, alg_dalg::CoDual; kwargs...)
+    # compute primal
+    D_ = Mooncake.primal(D_dD)
+    dD_ = Mooncake.tangent(D_dD)
+    A_ = Mooncake.primal(A_dA)
+    dA_ = Mooncake.tangent(A_dA)
+    A, dA = arrayify(A_, dA_)
+    D, dD = arrayify(D_, dD_)
+    Ac = copy(A)
+    Dc = copy(D)
+    dDc = copy(dD)
+    # update primal
+    DV = eig_full(A, Mooncake.primal(alg_dalg); kwargs...)
+    V = DV[2]
+    eig_vals!(A, D, Mooncake.primal(alg_dalg))
+    function deig_vals_adjoint(::Mooncake.NoRData)
+        dA .= zero(eltype(dA))
+        A .= Ac
+        PΔV = V' \ Diagonal(dD)
+        if eltype(dA) <: Real
+            ΔAc = PΔV * V'
+            dA .+= real.(ΔAc)
+        else
+            mul!(dA, PΔV, V', 1, 0)
+        end
```

Comment:
I think this logic can be achieved by passing an appropriate zerotangent for `V` to the `eig_pullback` function, see https://github.com/QuantumKitHub/MatrixAlgebraKit.jl/blob/c52614bffe120e3e63fbbd222230cdc695f476ea/src/pullbacks/eig.jl#L38

Reply:
Or you'd have to check that each element of the tangent is zero valued which could get expensive